sled-agent/src/support_bundle.rs

```diff
@@ -0,0 +1,291 @@
+    #[error("Failed to proccess output for command [{command}]: {error}")]
+    Output { command: String, error: std::io::Error },
+    #[error("Failed to create pipe for command [{command}]: {error}")]
+    Pipe { command: String, error: std::io::Error },
+    #[error("Failed to spawn command [{command}]: {error}")]
+    Spawn { command: String, error: std::io::Error },
+    #[error("Failed to execute command [{command}] in {duration:?}")]
+    Timeout { command: String, duration: Duration },
+    #[error("Failed to wait on command [{command}]: {error}")]
+    Wait { command: String, error: std::io::Error },
+}
+
+#[derive(Debug)]
+pub struct SupportBundleCmdOutput {
+    pub command: String,
+    pub stdio: String,
+    pub exit_status: String,
+}
+
+impl std::fmt::Display for SupportBundleCmdOutput {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        writeln!(f, "Command executed [{}]:", self.command)?;
+        writeln!(f, "    ==== stdio ====\n{}", self.stdio)?;
+        writeln!(f, "    ==== exit status ====\n{}", self.exit_status)
+    }
+}
+
+impl SupportBundleCommandHttpOutput
+    for Result<SupportBundleCmdOutput, SupportBundleCmdError>
+{
+    fn get_output(self) -> String {
+        match self {
+            Ok(output) => format!("{output}"),
+            Err(error) => format!("{error}"),
+        }
+    }
+}
+
+/// Takes a given `Command` and returns a lossy representation of the program
+// and it's arguments.
+fn command_to_string(command: &Command) -> String {
+    // Grab the command arguments
+    let mut res = command
+        .get_args()
+        .map(|a| a.to_string_lossy().into())
+        .collect::<Vec<String>>();
+
+    // Grab the command itself
+    res.insert(0, command.get_program().to_string_lossy().into());
+    res.join(" ")
+}
+
+/// Spawn a command asynchronously and collect it's output by interleaving
+/// stdout and stderr as they occur.
+async fn execute(
+    cmd: Command,
+    oneshot: oneshot::Sender<()>,
+) -> Result<SupportBundleCmdOutput, SupportBundleCmdError> {
+    let cmd_string = command_to_string(&cmd);
+    let (mut rx, tx) = os_pipe::pipe().map_err(|e| {
+        SupportBundleCmdError::Pipe { command: cmd_string.clone(), error: e }
+    })?;
+    let tx_clone = tx.try_clone().map_err(|e| SupportBundleCmdError::Dup {
+        command: cmd_string.clone(),
+        error: e,
+    })?;
+
+    // TODO MTZ: We may eventually want to reuse some of the process contract
+    // bits from illumos_utils to wrap the command in.
+    let mut cmd = tokio::process::Command::from(cmd);
+    cmd.kill_on_drop(true);
+    cmd.stdout(tx);
+    cmd.stderr(tx_clone);
+
+    let mut child = cmd.spawn().map_err(|e| SupportBundleCmdError::Spawn {
+        command: cmd_string.clone(),
+        error: e,
+    })?;
+    drop(cmd);
+
+    let output = tokio::task::spawn_blocking(
+        move || -> Result<String, std::io::Error> {
+            let mut out = String::new();
```

Comment:
Looks like this is blocking because `rx` exposes a non-async read interface.

WDYT about using https://docs.rs/tokio/latest/tokio/net/unix/pipe/fn.pipe.html instead? Then we'd be able to use a normal `tokio::task`, and we could just rely on it getting cancelled if we drop the future.

Reply:
Just another thought here, but let me see if I can wire something up that does `into_nonblocking_fd` + clone + `from_owned_fd`

Reply:
Yeah, I think "cloning" the sender using the raw FD is worth a shot. It might also be worth an upstream feature request, because I feel like being able to pass the same pipe into both a process's stdout and stderr is a fairly common use?

Reply:
...oh, you figured that bit out while i was writing the rest of my review. whoopsie! :)